plugins/woocommerce-admin/client/products/fields/attribute-field/attribute-field.tsx

```diff
@@ -15,29 +20,108 @@ import { closeSmall } from '@wordpress/icons';
 import './attribute-field.scss';
 import AttributeEmptyStateLogo from './attribute-empty-state-logo.svg';
 import { AddAttributeModal } from './add-attribute-modal';
+import { EditAttributeModal } from './edit-attribute-modal';
 import { reorderSortableProductAttributePositions } from './utils';
 
 type AttributeFieldProps = {
 	value: ProductAttribute[];
 	onChange: ( value: ProductAttribute[] ) => void;
+	productId?: number;
+};
+
+export type HydratedAttributeType = Omit< ProductAttribute, 'options' > & {
+	terms: ProductAttributeTerm[];
 };
 
 export const AttributeField: React.FC< AttributeFieldProps > = ( {
 	value,
 	onChange,
+	productId,
 } ) => {
 	const [ showAddAttributeModal, setShowAddAttributeModal ] =
 		useState( false );
+	const [ hydrationComplete, setHydrationComplete ] = useState< boolean >(
+		value ? false : true
+	);
+	const [ hydratedAttributes, setHydratedAttributes ] = useState<
+		HydratedAttributeType[]
+	>( [] );
+	const [ editingAttributeId, setEditingAttributeId ] = useState<
+		null | number
+	>( null );
+
+	const fetchTerms = useCallback(
+		( attributeId: number ) => {
+			return resolveSelect(
+				EXPERIMENTAL_PRODUCT_ATTRIBUTE_TERMS_STORE_NAME
+			)
+				.getProductAttributeTerms< ProductAttributeTerm[] >( {
+					attribute_id: attributeId,
+					product: productId,
+				} )
+				.then(
+					( attributeTerms ) => {
+						return attributeTerms;
+					},
+					( error ) => {
+						return error;
+					}
+				);
+		},
+		[ productId ]
+	);
+
+	useEffect( () => {
+		if ( ! value || hydrationComplete ) {
+			return;
+		}
+		Promise.all( value.map( ( attr ) => fetchTerms( attr.id ) ) ).then(
+			( allResults ) => {
+				setHydratedAttributes(
+					value.map( ( attr, index ) => {
+						const newAttr = {
+							...attr,
+							terms: allResults[ index ],
+							options: undefined,
+						};
+
+						return newAttr;
+					} )
+				);
+				setHydrationComplete( true );
+			}
+		);
+	}, [ productId, value, hydrationComplete ] );
+
+	useEffect( () => {
+		if ( ! hydrationComplete ) {
+			return;
+		}
+		onChange(
+			hydratedAttributes.map( ( attr ) => {
+				return {
+					...attr,
+					options: attr.terms.map( ( term ) => term.name ),
+					terms: undefined,
+				};
+			} )
+		);
```

Comment:
This also gets run after the `hydratedAttributes` are initial fetched, which makes the form dirty. Should we just set this as it's own function and only call it when the user updates or removes it?

Reply:
Well, the intention here was just to have the `hydratedAttributes` act as a proxy for the attributes in the form data, only with the `terms` data populated. So we can just alter this variable as we would otherwise alter the form data and get the behavior we want. It feels like it would complicate it more to have another callback that may or may not be called in conjunction with `onChange`.

Can you elaborate on what state would cause the form to become "dirty?" We'd only be updating this when we want to update the form data.

Reply:
> Can you elaborate on what state would cause the form to become "dirty?" We'd only be updating this when we want to update the form data.

You can notice this when looking at the `Update` button, when you load an existing product with attributes the `Update` button shows as disabled initially (meaning there are no changes) and once the `onChange` here is called it becomes active, allowing a user to save. This could also cause issues in the future if we decide to rely more on the `fieldsChanged` values.

I understand the reasoning behind this and see how you need it, but it be nice if we only trigger `onChange` once actual changes have been made and not just for the bootstrapping of the field.

The reason for suggesting a separate callback function is because this may be difficult to handle within a `useEffect`, as it would be difficult to differentiate if `hydratedAttributes` was updated by an actual change or just the initial loading.